### PR TITLE
refactor(gateway): avoid mutation in generators

### PR DESCRIPTION
**Pull request by michaelbeaumont**
### Summary

Previously there was a "hold and mutate everything object", making the logic and flow of data hard to follow.

### CODE REVIEW

File: pkg/plugins/runtime/gateway/generator.go [hunk 1]
```diff
@@ -159,41 +159,63 @@ func (g Generator) Generate(ctx xds_context.Context, proxy *core_xds.Proxy) (*co
 			Listener:         listener,
 		}
 
+		// This is checked by the gateway validator
+		if !SupportsProtocol(listener.Protocol) {
```
Comment (michaelbeaumont) [inline pkg/plugins/runtime/gateway/generator.go:163]: No protocols except HTTP/HTTPS are supported

**Review by jakubdyszkiewicz — APPROVED** [review]
